dashboard/app/models/ability.rb

```diff
@@ -74,8 +76,12 @@ def initialize(user)
       can? :update, level
     end
 
-    can [:show_by_keys], ProgrammingExpression do |expression|
-      can? :read, expression
+    can [:read], ProgrammingEnvironment do |environment|
+      environment.published || user.permission?(UserPermission::LEVELBUILDER)
+    end
+
+    can [:read, :show_by_keys], ProgrammingExpression do |expression|
+      can? :read, expression.programming_environment
```

Comment:
This logic is fairly straightforward but do we want to add some unit tests to ability_test.rb?

Reply:
Oh good call! The `ProgrammingEnvironment` logic is mostly covered by new tests in `programming_environments_controller_test` but I'll add a more comprehensive test